src/java.base/share/classes/jdk/internal/math/ToDecimal.java

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright (c) 2003, 2024, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2024, Alibaba Group Holding Limited. All Rights Reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package jdk.internal.math;
+
+import jdk.internal.access.JavaLangAccess;
+import jdk.internal.access.SharedSecrets;
+
+import static java.lang.Math.multiplyHigh;
+
+sealed class ToDecimal permits DoubleToDecimal, FloatToDecimal {
+    private static final JavaLangAccess JLA = SharedSecrets.getJavaLangAccess();
+
+    /* Used for left-to-tight digit extraction */
+    static final int MASK_28 = (1 << 28) - 1;
+
+    static final int NON_SPECIAL = 0;
+    static final int PLUS_ZERO   = 0x100;
+    static final int MINUS_ZERO  = 0x200;
+    static final int PLUS_INF    = 0x300;
+    static final int MINUS_INF   = 0x400;
+    static final int NAN         = 0x500;
+
+    static final byte LATIN1 = 0;
+    static final byte UTF16  = 1;
+
+    /**
+     * The identifier of the encoding used to encode the bytes. The supported values in this implementation are
+     *
+     * LATIN1
+     * UTF16
+     *
+     */
+    private final byte coder;
+
+    ToDecimal(byte coder) {
+        this.coder = coder;
+    }
+
+    final void putChar(byte[] str, int index, int c) {
```

Comment:
I think it would be better for all the `put...` methods here to return the updated `index`, hence replacing the return type from `void` to `int`.

This would make usages easier, as there's no need to remember at the call sites by how much the index must be incremented after the usage.